recbole/model/general_recommender/deosgcf.py

```diff
@@ -0,0 +1,213 @@
+    return scipy_to_sparse_tensor(get_symmetric_normalized(A))
+
+
+
+class SparseDropout(torch.nn.Module):
+    """
+    This is a Module that execute Dropout on Pytorch sparse tensor.
+    """
+
+    def __init__(self, p=0.5):
+        super().__init__()
+        # p is ratio of dropout
+        # convert to keep probability
+        self.kprob = 1 - p
+
+    def forward(self, x):
+        if not self.training:
+            return x
+
+        mask = ((torch.rand(x._values().size()) +
+                 self.kprob).floor()).type(torch.bool)
+        rc = x._indices()[:, mask]
+        val = x._values()[mask] * (1.0 / self.kprob)
+        return torch.sparse.FloatTensor(rc, val, x.shape)
+
+
+class DeosGCF(GeneralRecommender):
+    input_type = InputType.PAIRWISE
+
+    def __init__(self, config, dataset):
+        super().__init__(config, dataset)
+
+        # load dataset info
+        self.interaction_matrix = dataset.inter_matrix(
+            form='coo').astype(np.float32)
+
+        # load parameters info
+        self.latent_dim = config['embedding_size']
+        self.n_layers = config['n_layers']
+        self.reg_weight = config['reg_weight']
+        self.dropout_prob = config['dropout_prob']
+
+        # define layers and loss
+        self.user_embedding = torch.nn.Embedding(
+            num_embeddings=self.n_users, embedding_dim=self.latent_dim)
+        self.item_embedding = torch.nn.Embedding(
+            num_embeddings=self.n_items, embedding_dim=self.latent_dim)
+
+        # will get a warning about this ParameterList ".training" attr getting set until this goes through
+        # https://github.com/pytorch/pytorch/pull/48315
+        self.la_layer_params = torch.nn.ParameterList([torch.nn.Parameter(
+            torch.ones(self.n_users + self.n_items)) for _ in range(self.n_layers)])
+        
+        # xavier init the LA layer params
+        # not sure this even makes sense, but imitating the existing TF implementation
+        # https://github.com/JimLiu96/DeosciRec/blob/b3575da96908d062fcf23a0d4fd5f3e3f082573d/DGCF_osci.py#L146
+        # since there is no defined "fanout" have to do this manually
+        # i.e., "Fan in and fan out can not be computed for tensor with fewer than 2 dimensions"
+        uniform_range = np.sqrt(6/(self.n_users+self.n_items+1))
+        for weights in self.la_layer_params:
+            torch.nn.init.uniform_(weights, -uniform_range, uniform_range)
+
+        self.sparse_dropout = SparseDropout(p=self.dropout_prob)
+        
+        self.mf_loss = BPRLoss()
+        self.reg_loss = EmbLoss()
+
+        # storage variables for full sort evaluation acceleration
+        self.restore_user_e = None
+        self.restore_item_e = None
+
+        # generate intermediate data
+        A = adjacency_of_bipartite(self.interaction_matrix)
+        norm_adj_matrix = get_symm_norm_tensor(A).to(self.device)
+        norm_crosshop_matrix = get_symm_norm_tensor(A**2).to(self.device)
```

Comment:
There is a high-pass filtering on the corss-hop metrix which greatly improves the efficiency. It should also be implemented.
![paper1](https://user-images.githubusercontent.com/29095067/125239827-cf82a580-e31b-11eb-9f47-77076f1bfd34.jpg)